velox/functions/prestosql/window/Rank.cpp

```diff
@@ -41,17 +51,49 @@ class RankFunction : public exec::WindowFunction {
       vector_size_t resultOffset,
       const VectorPtr& result) override {
     int numRows = peerGroupStarts->size() / sizeof(vector_size_t);
-    auto* rawValues = result->asFlatVector<int64_t>()->mutableRawValues();
     auto* rawPeerStarts = peerGroupStarts->as<vector_size_t>();
+
+    void* rawValues = nullptr;
+    switch (type) {
+      case RankType::kRank:
+      case RankType::kDenseRank: {
+        rawValues = result->asFlatVector<int64_t>()->mutableRawValues();
+        break;
+      }
+      case RankType::kPercentRank: {
+        rawValues = result->asFlatVector<double>()->mutableRawValues();
+        break;
+      }
+    }
+
     for (int i = 0; i < numRows; i++) {
       auto start = rawPeerStarts[i];
       if (start != currentPeerGroupStart_) {
         currentPeerGroupStart_ = start;
-        rank_ += previousPeerCount_;
+        if constexpr (type == RankType::kDenseRank) {
+          rank_ += (previousPeerCount_ != 0);
+        } else {
+          rank_ += previousPeerCount_;
+        }
         previousPeerCount_ = 0;
       }
 
-      rawValues[resultOffset + i] = rank_;
+      switch (type) {
+        case RankType::kRank: {
+          ((int64_t*)rawValues)[resultOffset + i] = rank_;
```

Comment:
Casting on every row is also expensive.

Reply:
Thanks for the input, since the vector result is passed only on the call to ``apply()``, I am not sure how rawValues can be accessed and its type cast beforehand. Hence a void pointer was used here. Is there any alternative to using a void pointer here, other than maybe splitting it into three separate classes?

Reply:
Perhaps, add 2nd template parameter to the class for the result type.

Reply:
Thanks for the suggestion, changed accordingly!